docs: Blogpost about improving time to interactive in gatsby-plugin-segment-js (#17179)
gatsbyjs/gatsby: docs/blog/2019-08-30-speed-up-your-time-to-interactive-by-delaying-third-party-scripts/index.md

---
title: Speed up your time to interactive by delaying third party scripts
date: 2019-08-30
author: Kilian Valkhof
excerpt: "Delay loading scripts to lower your time to interactive by over 60%"
tags:
  - performance
  - seo
---

Gatsby does many things to get a website visible as fast as possible. But getting a website visible isn't the only interesting metric. There are others that, if you're not careful, can easily take much longer than needed. Even if your site still feels fast.

The one this article focuses on is "[Time to interactive](https://developers.google.com/web/tools/lighthouse/audits/time-to-interactive)", which is the time it takes for all your initial scripts to run. This one is particularly interesting for a couple of reasons:

> - It's not super noticable when you use a website on desktop

Reviewer: ```suggestion
- It's not super noticeable when you use a website on desktop
```

- async scripts are counted towards this number (Even though they're async, they're still loaded in sequence after a page load)
- It's a very important part of [Google's performance ranking](https://github.com/GoogleChrome/lighthouse/blob/master/docs/scoring.md). It's 5 times more important than the "[first meaningful paint](https://developers.google.com/web/tools/lighthouse/audits/first-meaningful-paint)", e.g. when something shows up on the screen. **5 times!**

> These together might mean that you have a website that _feels_ fast, but Google will give you a penalty because it's slow.

Author: The "feels" refers to it being visible and rendered. If you go from a blank page to a rendering page in a couple of seconds without reflows, that is what makes it feel fast to me. At that point scrolling/hovers etc work, which is what most people will do in the first couple of seconds. From what I've seen, Google TTI measure is something of a worst-case scenario number.

> To get an overview of how your site scores, you can use Lighthouse, either in your Chrome developer tools by going to the "Audits" tab, or by going to [web.dev](https://web.dev/measure). This will test your site on a number of different metrics, such as accessibility, SEO (Search Engine Optimization), and best practices, but the one we're interested in is "performance".

Reviewer: there is also https://developers.google.com/speed/pagespeed/insights/ if you just want performance or the chrome extension if you want the latest and greatest lighthouse. Audit panel is mostly on an outdated version.

I don't think it's necessary to point this out but just stating it in the review as other possible values.

Author: Yeah there's the CLI too etc. I think pointing people to web.dev is the best method but that's not available for all sites (localhosts, intranet pages), then the audit panel is better.

> To get an overview of how your site scores, you can use Lighthouse, either in your Chrome developer tools by going to the "Audits" tab, or by going to [web.dev](https://web.dev/measure). This will test your site on a number of different metrics, such as accessibility, SEO (Search Engine Optimization), and best practices, but the one we're interested in is "performance".

Reviewer: ```suggestion
To get an overview of how your site scores, you can use Lighthouse, either in your Chrome developer tools by going to the "Audits" tab, or by going to [web.dev](https://web.dev/measure). Lighthouse will test your site on a number of different metrics, such as accessibility, SEO (Search Engine Optimization), and best practices, but the one we're interested in is "performance".
```

Harry from [marketingexamples.com](https://marketingexamples.com/) linked me to a recent post of his on [SEO performance](https://marketingexamples.com/seo/performance) that mentioned the "Time to interactive" scoring and I decided to take another look at my website for [Polypane](https://polypane.rocks). I had checked it in the past and it had a good (90+) score, but I was pretty shocked when it came back with a suddenly had a score of 63!

When looking at the performance data, the site's first meaningful paint took slightly over a second, but the time to interactive was _eleven seconds_. Yikes!

The reason for this was that I recently switched to [Segment](https://segment.com) (using [gatsby-plugin-segment-js](//packages/gatsby-plugin-segment-js/)) and was loading other scripts through that, like support chat and analytics. These scripts all counted towards my "time to interactive".

The SEO performance post included a tip from Dave of [ToDesktop](https://www.todesktop.com/) who has similar problems. His tip: Prevent loading the scripts until after a user has scrolled, along with some timeout to prevent [scroll jank](http://jankfree.org/).

> This way, your "Time to interactive" won't take these scripts into account, and the user won't need your support widget in the first second of a page anyway, so this works well for everyone. There is a downside: your [bounce rate](https://support.google.com/analytics/answer/1009409?hl=en) will become less accurate as the people that open your site, don't scroll and leave will never show up in your analytics.

Reviewer: ```suggestion
By adding a timeout, your "Time to interactive" won't take these scripts into account. The user won't need your support widget in the first second of a page anyway, so this works well for everyone.. With any upside, there is always a downside: your [bounce rate](https://support.google.com/analytics/answer/1009409?hl=en) will become less accurate as the people that open your site, don't scroll and leave will never show up in your analytics.
```

To work around this, I added two locks to the Gatsby plugin:

- `segmentSnippetLoaded`, `false` by default and set to `true` after it's loaded.
- `segmentSnippetLoading`, `true` only between when the load function has been called and when it has finished.

Then, either on scroll or onRouteChange, we only call the load function if segmentSnippetLoaded is not true, and in the load function, we only continue if segmentSnippetLoading is not true. This prevents the function from being called at all after the first time the script has been loaded. If the function is called twice but we're still in the countdown time, nothing happens.

After implementing this on my own website, my performance score shot back up from 63 to 94 and I had an over 60% decrease in the time to interactive. Pretty good for a few lines of code.

There is currently a [PR](https://github.com/benjaminhoffman/gatsby-plugin-segment-js/pull/19) open to add a `delayLoad` option to gatsby-plugin-segment-js to enable this feature. Alternatively, you can build it from [my fork](https://github.com/Kilian/gatsby-plugin-segment-js).

## In closing

> If you work on a laptop or desktop with a good internet connection, the time to interactive is not so noticable as "time before you see something on your screen", but it can have a big impact on your performance rating and because of that on your SEO. You should be extra careful when adding third party scripts to your page, and the best way to do that is to run Google Lighthouse. If you notice your Gatsby site has a bad performance score, make sure to check the Time to interactive value. If it's very high, try loading scripts only after a user has interacted with your website.

Reviewer 2: ```suggestion
If you work on a laptop or desktop with a good internet connection, the Time to Interactive is not so noticable as "time before you see something on your screen", but it can have a big impact on your performance rating and interactivity, and because of that, your SEO. You should be extra careful when adding third party scripts to your page, and the best way to do that is to run Google Lighthouse. If you notice your Gatsby site has a bad performance score, make sure to check the Time to Interactive value. If it's very high, try loading scripts only after a user has interacted with your website.
```
